### PR TITLE
Ticket25762 034 01

### PR DESCRIPTION
**Pull request by dgoulet-tor**
Some stuff are still missing but this is a first pass at this. See ticket on trac for details.

### CODE REVIEW

File: src/or/main.c [hunk 1]
```diff
@@ -1458,6 +1458,55 @@ initialize_periodic_events_cb(evutil_socket_t fd, short events, void *data)
   }
 }
 
+/* Helper function: Set up periodic_events[] that only matches the given
```
Comment (nmathewson) [inline src/or/main.c]: Documentation should be a doxygen comment, and should explain what "matches" means.

File: src/or/main.c [hunk 2]
```diff
@@ -1458,6 +1458,55 @@ initialize_periodic_events_cb(evutil_socket_t fd, short events, void *data)
...
+ * roles. */
+static void
+setup_periodic_events_by_roles(uint32_t roles)
+{
+  for (int i = 0; periodic_events[i].name; ++i) {
+    periodic_event_item_t *item = &periodic_events[i];
+    if (item->roles & roles) {
+      periodic_event_setup(item);
```
Comment (nmathewson) [inline src/or/main.c]: How does this work when we call this function later on?  If we're going to do this, we need to make periodic_event_setup idempotent.

Actually, we can safely just call periodic_event_setup() on all of these events, and use periodic_event_launch to start the ones we really want. We can add a new periodic_event_disable() to turn off the ones we don't want.

File: src/or/main.c [hunk 3]
```diff
@@ -1458,6 +1458,55 @@ initialize_periodic_events_cb(evutil_socket_t fd, short events, void *data)
...
+
+/* Helper function: Set up all events in periodic_events[] and filter them by
+ * roles for which we are configured for. */
+static void
+setup_periodic_events(void)
+{
+  const or_options_t *options = get_options();
+
+  /* At this stage, we are booting up so we need to know for which roles this
+   * tor instance has. */
```
Comment (nmathewson) [inline src/or/main.c]: Let's make just one bitfield here, and call setup_periodic_events_by_roles once?

File: src/or/main.c [hunk 4]
```diff
@@ -1551,6 +1551,21 @@ teardown_periodic_events(void)
...
+ * periodic events. */
+void
+periodic_events_on_new_options(void)
+{
+  /* Only if we've already initialized once the events, teardown them all and
+   * reinitialize. It is just simpler that way instead of going through all
+   * currently enabled events and trying to destroy only the one that could be
+   * affected. */
+  if (periodic_events_initialized) {
+    teardown_periodic_events();
```
Comment (nmathewson) [inline src/or/main.c]: This is not okay -- if we tear them down and then set them up again, the  actual timers will all be reset.

**Comment by dgoulet-tor** [comment]
This was merged with another PR.

**Pull request closed without merging**